numpy/ma/core.py

```diff
@@ -3109,8 +3109,6 @@ def __setitem__(self, indx, value):
                 _mask[indx] = tuple([True] * nbfields)
             else:
                 _mask[indx] = True
-            if not self._isfield:
```

Comment:
I'll start by explaining what is going on and then answer your questions if that is alright. Please forgive me if I go into too much detail and explain the obvious. Good questions by the way. Sorry if my answer gets too long.

This particular branch is dealing with the case of a structured array. Suppose for example we have a structured array created like so `a = np.zeros((2,), dtype=[("c",int, 2), ("d", float, 3)])`. We could construct a masked structured array like so `b = np.ma.array(a, mask=np.ma.getmaskarray(a))`. In this example, we have two fields `c` and `d`. If we index into one say `c` like so `b["c"]`, we will get a normal masked array.

However, there is no way to know that this masked array came from a structured masked array; so, the `_isfield` member variable indicates this. One can verify this is the case by trying `b._isfield` and `b["c"]._isfield`.

Now, why was `_isfield` checked? My belief, as I can't speak for the author personally, is the author correctly assumed that if we are looking at a field of a structured array our mask is a view of the mask from the original structured array's mask and we want to ensure changes here are propagated back to the original and so this must remain shared if it already is (this indirectly answers your first question). However, if it is not shared, it will remain the case.

To directly answer your first question, we can't assume anything about `self._sharedmask`. Its value will be dependent on what .

The argument that I am making here is that this one can't assume this is not a view if it is not a field. In other words, we could always be working with a view of another masked array where we want changes propagated to an original and should not be forcing the user's hand here. Additionally, it would be inconsistent to "unshare" the mask as the data remains shared regardless.